clippy_lints/src/swap.rs

```diff
@@ -262,3 +264,82 @@ fn check_suspicious_swap(cx: &LateContext<'_>, block: &Block<'_>) {
         }
     }
 }
+
+/// Implementation of the xor case for `MANUAL_SWAP` lint.
+fn check_xor_swap(cx: &LateContext<'_>, block: &Block<'_>) {
+    for window in block.stmts.windows(3) {
+        if_chain! {
+            if let Some((lhs0, rhs0)) = extract_sides_of_xor_assign(&window[0]);
+            if let Some((lhs1, rhs1)) = extract_sides_of_xor_assign(&window[1]);
+            if let Some((lhs2, rhs2)) = extract_sides_of_xor_assign(&window[2]);
+            if eq_expr_value(cx, lhs0, rhs1);
+            if eq_expr_value(cx, rhs1, lhs2);
+            if eq_expr_value(cx, rhs0, lhs1);
+            if eq_expr_value(cx, lhs1, rhs2);
+            then {
+                let span = window[0].span.to(window[2].span);
+                let mut applicability = Applicability::MachineApplicable;
+                match check_for_slice(cx, lhs0, rhs0) {
```

Comment:
`check_for_slice` was there before this PR. So I don't think it is useless. And your changes seems doesn't detect vector and array:
```
let foo = &mut [1, 2];
-    foo.swap(0, 1)
+    std::mem::swap(&mut foo[0], &mut foo[1])
```

Reply:
> `check_for_slice` was there before this PR. So I don't think it is useless.

Well I didn't say it was useless. But that's okay to leave it.

> And your changes seems doesn't detect vector and array:

Ah my bad. I thought the indexing operation always derefs to slice but that is not the case.

Reply:
I found a false negative in old code:
```rust
fn distinct_slice() {
    let foo = &mut [vec![1, 2], vec![3, 4]];
    let bar = &mut [vec![1, 2], vec![3, 4]];
    let temp = foo[0][1];
    foo[0][1] = bar[1][0];
    bar[1][0] = temp;
}
```
So now I'm with you on removing `check_for_slice`. Is there a function that checks if we can mutably borrow two expression at the same time? It will also solve fields problem.

Reply:
Oof no I'm afraid not. I think you would have to manually traverse the expression.

If you are swapping `foo.a...` and `foo.b...` that is okay because `a` and `b` differ. But if you are swapping `foo.a[b]...` and `foo.a[c]...` that is not okay because indices differ. So its tricky because you traverse into the expression "right to left" but you'll have to compare them "left to right".

You don't have to fix the false negative in this PR but its up to you.

Reply:
I done it, but there may be a better way. Also I removed `check_for_slice`.